quickjs.c

```diff
@@ -50132,3 +50417,168 @@ void JS_AddPerformance(JSContext *ctx)
                            JS_PROP_ENUMERABLE);
     JS_FreeValue(ctx, performance);
 }
+
+/* Poly IC */
+
+static force_inline uint32_t get_index_hash(JSAtom atom, int hash_bits)
+{
+    return (atom * 0x9e370001) >> (32 - hash_bits);
+}
+
+InlineCache *init_ic(JSContext *ctx)
+{
+    InlineCache *ic;
+    ic = js_malloc(ctx, sizeof(InlineCache));
+    if (unlikely(!ic))
+        goto fail;
+    ic->count = 0;
+    ic->hash_bits = 2;
+    ic->capacity = 1 << ic->hash_bits;
+    ic->ctx = ctx;
+    ic->hash = js_malloc(ctx, sizeof(ic->hash[0]) * ic->capacity);
+    if (unlikely(!ic->hash))
+        goto fail;
+    memset(ic->hash, 0, sizeof(ic->hash[0]) * ic->capacity);
+    ic->cache = NULL;
+    ic->updated = FALSE;
+    ic->updated_offset = 0;
+    return ic;
+fail:
+    return NULL;
```

Comment:
```suggestion
    js_free(ctx, ic);
    return NULL;
```
Leaks memory otherwise. The first `goto fail` could be a `return NULL`.